src/Symfony/Component/Messenger/Transport/Doctrine/Connection.php

```
@@ -0,0 +1,232 @@
+     * Configuration of the connection.
+     *
+     * * table_name : name of the table
+     * * connection : name of the Doctrine's entity manager
+     * * redeliver_timeout : Timeout before redeliver messages still in handling state (i.e: delivered_at is not null and message is still in table). Default 3600
+     * * loop_sleep : Number of micro seconds to wait for a next message to handle
+     */
+    private $configuration = [];
+    private $driverConnection;
+
+    public function __construct(array $configuration, DriverConnection $driverConnection)
+    {
+        $this->configuration = array_replace_recursive([
+            'table_name' => 'messenger_messages',
+            'queue' => 'default',
+            'redeliver_timeout' => 3600,
+            'loop_sleep' => 200000,
+            'auto_setup' => true,
+        ], $configuration);
+        $this->driverConnection = $driverConnection;
+    }
+
+    public function getConfiguration(): array
+    {
+        return $this->configuration;
+    }
+
+    public static function buildConfiguration($dsn, array $options = [])
+    {
+        if (false === $parsedUrl = parse_url($dsn)) {
+            throw new InvalidArgumentException(sprintf('The given Doctrine DSN "%s" is invalid.', $dsn));
+        }
+
+        $components = parse_url($dsn);
+        $query = [];
+        if (isset($components['query'])) {
+            parse_str($components['query'], $query);
+        }
+
+        return [
+            'connection' => $components['host'],
+            'table_name' => $options['table_name'] ?? ($query['table_name'] ?? 'messenger_messages'),
+            'queue' => $options['queue'] ?? ($query['queue'] ?? 'default'),
+            'redeliver_timeout' => $options['redeliver_timeout'] ?? ($query['redeliver_timeout'] ?? 3600),
+            'loop_sleep' => $options['loop_sleep'] ?? ($query['loop_sleep'] ?? 200000),
+            'auto_setup' => $options['auto_setup'] ?? ($query['auto_setup'] ?? true),
+        ];
```

We haven't done this in the Amqp Connection (yet), but I think we should validate the `$options` and `$query`. If something we don't recognize is passed, we should throw an exception and list what the valid options *are*.

---

Can we use the OptionsResolver component ? or we must avoid a new dependency ?

---

> If there are extra keys, throw an exception.

Let's do this 👍